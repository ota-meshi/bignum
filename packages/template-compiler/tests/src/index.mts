import { compile } from "../../src/index.mjs";
import type { BTContext } from "@bignum/shared";
import * as snap from "@ota-meshi/test-snapshot";

const context: BTContext<number | bigint | string, string> = {
  binaryOperations: {
    "+": (a, b) => `(${a} + ${b})`,
    "-": (a, b) => `(${a} - ${b})`,
    "*": (a, b) => `(${a} * ${b})`,
    "/": (a, b) => `(${a} / ${b})`,
    "%": (a, b) => `(${a} % ${b})`,
    "**": (a, b) => `(${a} ** ${b})`,
    "<": (a, b) => `(${a} < ${b})`,
    "<=": (a, b) => `(${a} <= ${b})`,
    ">": (a, b) => `(${a} > ${b})`,
    ">=": (a, b) => `(${a} >= ${b})`,
    "==": (a, b) => `(${a} == ${b})`,
    "!=": (a, b) => `(${a} != ${b})`,
  },
  unaryOperations: {
    "-": (a) => `(-${a})`,
    "+": (a) => `(+${a})`,
  },
  variables: {
    PI: "PI",
  },
  functions: {
    abs: (a) => `abs(${a})`,
    trunc: (a) => `trunc(${a})`,
    round: (a) => `round(${a})`,
    ceil: (a) => `ceil(${a})`,
    floor: (a) => `floor(${a})`,
    foo: (...args) => `foo(${args.join(", ")})`,
  },
};

describe("standard tests", () => {
  for (const t of [
    test`1 + 2`,
    test`1+2`,
    test`1.1 + 2.1`,
    test`1+-2`,
    test`1-+2`,
    test`1.1+-2.2`,
    test`1.2+0.8`,
    test`1 + ${3}`,
    test`1 + ${2} * 3`,
    test`1.2 + ${2.3} * 3.4`,
    test`(1 + ${2}) * 3`,
    test`1 * -(23 % 4)`,
    test`1.5 * -(23 % 4)`,
    test`3*3*PI`,
    test`3.2*3.2*PI`,
    test`trunc(300*1.08)`,
    test`trunc(320*1.08)`,
    test`1 + 2 * 3`,
    test`3 - 3 / 3`,
    test`3.2 - 3.2 / 3.2`,
    test`(1 + 2) * 3`,
    test`2 * 2 ** 3`,
    test`2.1 * 2.3 ** 3`,
    test`2.1 * 2.3 ** (1.5+1.5)`,
    // test`2.1 * 2.3 ** 3.4`, // Error
    test`4 / 2 ** 2`,
    test`4.1 / 2.2 ** 2`,
    test`(2 * 2) ** 3`,
    test`abs(-10)`,
    test`(2 * 2) / 10 ** 20`,
    test`(2 * 2) / 10 ** 21`, // Overflow the default maximum number of decimal places. (20).
    test`2.3 / 1.1`,
    test`2.3 % 1.1`,
    test`22000000 / 1.1`,
    ...[
      "1.49",
      "1.5",
      "1.51",
      "1.501",
      "-1.49",
      "-1.5",
      "-1.51",
      "-1.501",
    ].flatMap((v) => [
      test`trunc(${v})`,
      test`round(${v})`,
      test`floor(${v})`,
      test`ceil(${v})`,
      test`abs(${v})`,
    ]),
    test`1e42+1`,
    test`1e-42+1`,
    test`1e+42+1`,
    test`1.23e42+1`,
    test`1.23e-42+1`,
    test`1.23e+42+1`,
    test`-1.23e42+1`,
    test`-1.23e-42+1`,
    test`-1.23e+42+1`,
    test`foo() + foo(1) + foo(1, 2) + foo(1, 2, 3)`,
    test`2 ** 3 ** 4`,
    test`(2 ** 3) ** 4`,
  ]) {
    it(t.name, () => {
      snap
        .expect(compile(t.template)(t.substitutions, context))
        .toMatchSnapshot();
    });
  }
});
describe("Big Number tests", () => {
  for (const t of [
    test`0.1 + 0.1 + 0.1`,
    test`${0.1} + ${0.1} + ${0.1}`,
    test`0.2 + 0.1`,
    test`${0.2} + ${0.1}`,
    test`0.3 - 0.1`,
    test`${0.3} - ${0.1}`,
  ]) {
    it(t.name, () => {
      snap
        .expect(compile(t.template)(t.substitutions, context))
        .toMatchSnapshot();
    });
  }
});
describe("Compare tests", () => {
  for (const t of [
    test`0.1 + 0.1 + 0.1 <= 0.3`,
    test`${0.1} + ${0.1} + ${0.1} == ${0.3}`,
    test`0.2 + 0.1 != 0.3`,
    test`${0.2} + ${0.1} != ${0.3}`,
    test`0.3 - 0.1 >= 0.2`,
    test`${0.3} - ${0.1} >= ${0.2}`,
    test`${0.3} - ${0.1} < ${0.2}`,
    test`${0.3} - ${0.1} > ${0.2}`,
    ...[
      [0.1, 0.1],
      [0.1, 0.2],
      [0.2, 0.1],
    ].flatMap(([a, b]) => [
      test`${a} < ${b}`,
      test`${a} <= ${b}`,
      test`${a} > ${b}`,
      test`${a} >= ${b}`,
      test`${a} == ${b}`,
      test`${a} != ${b}`,
    ]),
  ]) {
    it(t.name, () => {
      snap
        .expect(compile(t.template)(t.substitutions, context))
        .toMatchSnapshot();
    });
  }
});

/**
 * Make test data
 */
function test(
  template: TemplateStringsArray,
  ...substitutions: (number | bigint | string)[]
) {
  let name = "`";
  for (let index = 0; index < template.length; index++) {
    name += template[index];
    if (index < substitutions.length) {
      name += `\${${substitutions[index]}}`;
    }
  }
  name += "`";
  return {
    name,
    template,
    substitutions,
  };
}
