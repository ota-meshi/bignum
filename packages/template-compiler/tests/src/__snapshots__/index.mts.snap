// test-snapshot v1

exports[`Big Number tests \`\${0.1} + \${0.1} + \${0.1}\` 1`] = `"((0.1 + 0.1) + 0.1)"`;

exports[`Big Number tests \`\${0.2} + \${0.1}\` 1`] = `"(0.2 + 0.1)"`;

exports[`Big Number tests \`\${0.3} - \${0.1}\` 1`] = `"(0.3 - 0.1)"`;

exports[`Big Number tests \`0.1 + 0.1 + 0.1\` 1`] = `"((0.1 + 0.1) + 0.1)"`;

exports[`Big Number tests \`0.2 + 0.1\` 1`] = `"(0.2 + 0.1)"`;

exports[`Big Number tests \`0.3 - 0.1\` 1`] = `"(0.3 - 0.1)"`;

exports[`Compare tests \`\${0.1} != \${0.1}\` 1`] = `"(0.1 != 0.1)"`;

exports[`Compare tests \`\${0.1} != \${0.2}\` 1`] = `"(0.1 != 0.2)"`;

exports[`Compare tests \`\${0.1} + \${0.1} + \${0.1} == \${0.3}\` 1`] = `"(((0.1 + 0.1) + 0.1) == 0.3)"`;

exports[`Compare tests \`\${0.1} < \${0.1}\` 1`] = `"(0.1 < 0.1)"`;

exports[`Compare tests \`\${0.1} < \${0.2}\` 1`] = `"(0.1 < 0.2)"`;

exports[`Compare tests \`\${0.1} <= \${0.1}\` 1`] = `"(0.1 <= 0.1)"`;

exports[`Compare tests \`\${0.1} <= \${0.2}\` 1`] = `"(0.1 <= 0.2)"`;

exports[`Compare tests \`\${0.1} == \${0.1}\` 1`] = `"(0.1 == 0.1)"`;

exports[`Compare tests \`\${0.1} == \${0.2}\` 1`] = `"(0.1 == 0.2)"`;

exports[`Compare tests \`\${0.1} > \${0.1}\` 1`] = `"(0.1 > 0.1)"`;

exports[`Compare tests \`\${0.1} > \${0.2}\` 1`] = `"(0.1 > 0.2)"`;

exports[`Compare tests \`\${0.1} >= \${0.1}\` 1`] = `"(0.1 >= 0.1)"`;

exports[`Compare tests \`\${0.1} >= \${0.2}\` 1`] = `"(0.1 >= 0.2)"`;

exports[`Compare tests \`\${0.2} != \${0.1}\` 1`] = `"(0.2 != 0.1)"`;

exports[`Compare tests \`\${0.2} + \${0.1} != \${0.3}\` 1`] = `"((0.2 + 0.1) != 0.3)"`;

exports[`Compare tests \`\${0.2} < \${0.1}\` 1`] = `"(0.2 < 0.1)"`;

exports[`Compare tests \`\${0.2} <= \${0.1}\` 1`] = `"(0.2 <= 0.1)"`;

exports[`Compare tests \`\${0.2} == \${0.1}\` 1`] = `"(0.2 == 0.1)"`;

exports[`Compare tests \`\${0.2} > \${0.1}\` 1`] = `"(0.2 > 0.1)"`;

exports[`Compare tests \`\${0.2} >= \${0.1}\` 1`] = `"(0.2 >= 0.1)"`;

exports[`Compare tests \`\${0.3} - \${0.1} < \${0.2}\` 1`] = `"((0.3 - 0.1) < 0.2)"`;

exports[`Compare tests \`\${0.3} - \${0.1} > \${0.2}\` 1`] = `"((0.3 - 0.1) > 0.2)"`;

exports[`Compare tests \`\${0.3} - \${0.1} >= \${0.2}\` 1`] = `"((0.3 - 0.1) >= 0.2)"`;

exports[`Compare tests \`0.1 + 0.1 + 0.1 <= 0.3\` 1`] = `"(((0.1 + 0.1) + 0.1) <= 0.3)"`;

exports[`Compare tests \`0.2 + 0.1 != 0.3\` 1`] = `"((0.2 + 0.1) != 0.3)"`;

exports[`Compare tests \`0.3 - 0.1 >= 0.2\` 1`] = `"((0.3 - 0.1) >= 0.2)"`;

exports[`standard tests \`(1 + \${2}) * 3\` 1`] = `"((1 + 2) * 3)"`;

exports[`standard tests \`(1 + 2) * 3\` 1`] = `"((1 + 2) * 3)"`;

exports[`standard tests \`(2 * 2) ** 3\` 1`] = `"((2 * 2) ** 3)"`;

exports[`standard tests \`(2 * 2) / 10 ** 20\` 1`] = `"((2 * 2) / (10 ** 20))"`;

exports[`standard tests \`(2 * 2) / 10 ** 21\` 1`] = `"((2 * 2) / (10 ** 21))"`;

exports[`standard tests \`(2 ** 3) ** 4\` 1`] = `"((2 ** 3) ** 4)"`;

exports[`standard tests \`-1.23e+42+1\` 1`] = `"((-1.23e+42) + 1)"`;

exports[`standard tests \`-1.23e-42+1\` 1`] = `"((-1.23e-42) + 1)"`;

exports[`standard tests \`-1.23e42+1\` 1`] = `"((-1.23e42) + 1)"`;

exports[`standard tests \`1 * -(23 % 4)\` 1`] = `"(1 * (-(23 % 4)))"`;

exports[`standard tests \`1 + \${2} * 3\` 1`] = `"(1 + (2 * 3))"`;

exports[`standard tests \`1 + \${3}\` 1`] = `"(1 + 3)"`;

exports[`standard tests \`1 + 2 * 3\` 1`] = `"(1 + (2 * 3))"`;

exports[`standard tests \`1 + 2\` 1`] = `"(1 + 2)"`;

exports[`standard tests \`1+-2\` 1`] = `"(1 + (-2))"`;

exports[`standard tests \`1+2\` 1`] = `"(1 + 2)"`;

exports[`standard tests \`1.1 + 2.1\` 1`] = `"(1.1 + 2.1)"`;

exports[`standard tests \`1.1+-2.2\` 1`] = `"(1.1 + (-2.2))"`;

exports[`standard tests \`1.2 + \${2.3} * 3.4\` 1`] = `"(1.2 + (2.3 * 3.4))"`;

exports[`standard tests \`1.2+0.8\` 1`] = `"(1.2 + 0.8)"`;

exports[`standard tests \`1.5 * -(23 % 4)\` 1`] = `"(1.5 * (-(23 % 4)))"`;

exports[`standard tests \`1.23e+42+1\` 1`] = `"(1.23e+42 + 1)"`;

exports[`standard tests \`1.23e-42+1\` 1`] = `"(1.23e-42 + 1)"`;

exports[`standard tests \`1.23e42+1\` 1`] = `"(1.23e42 + 1)"`;

exports[`standard tests \`1-+2\` 1`] = `"(1 - (+2))"`;

exports[`standard tests \`1e+42+1\` 1`] = `"(1e+42 + 1)"`;

exports[`standard tests \`1e-42+1\` 1`] = `"(1e-42 + 1)"`;

exports[`standard tests \`1e42+1\` 1`] = `"(1e42 + 1)"`;

exports[`standard tests \`2 * 2 ** 3\` 1`] = `"(2 * (2 ** 3))"`;

exports[`standard tests \`2 ** 3 ** 4\` 1`] = `"(2 ** (3 ** 4))"`;

exports[`standard tests \`2.1 * 2.3 ** (1.5+1.5)\` 1`] = `"(2.1 * (2.3 ** (1.5 + 1.5)))"`;

exports[`standard tests \`2.1 * 2.3 ** 3\` 1`] = `"(2.1 * (2.3 ** 3))"`;

exports[`standard tests \`2.3 % 1.1\` 1`] = `"(2.3 % 1.1)"`;

exports[`standard tests \`2.3 / 1.1\` 1`] = `"(2.3 / 1.1)"`;

exports[`standard tests \`3 - 3 / 3\` 1`] = `"(3 - (3 / 3))"`;

exports[`standard tests \`3*3*PI\` 1`] = `"((3 * 3) * PI)"`;

exports[`standard tests \`3.2 - 3.2 / 3.2\` 1`] = `"(3.2 - (3.2 / 3.2))"`;

exports[`standard tests \`3.2*3.2*PI\` 1`] = `"((3.2 * 3.2) * PI)"`;

exports[`standard tests \`4 / 2 ** 2\` 1`] = `"(4 / (2 ** 2))"`;

exports[`standard tests \`4.1 / 2.2 ** 2\` 1`] = `"(4.1 / (2.2 ** 2))"`;

exports[`standard tests \`22000000 / 1.1\` 1`] = `"(22000000 / 1.1)"`;

exports[`standard tests \`abs(\${-1.5})\` 1`] = `"abs(-1.5)"`;

exports[`standard tests \`abs(\${-1.49})\` 1`] = `"abs(-1.49)"`;

exports[`standard tests \`abs(\${-1.51})\` 1`] = `"abs(-1.51)"`;

exports[`standard tests \`abs(\${-1.501})\` 1`] = `"abs(-1.501)"`;

exports[`standard tests \`abs(\${1.5})\` 1`] = `"abs(1.5)"`;

exports[`standard tests \`abs(\${1.49})\` 1`] = `"abs(1.49)"`;

exports[`standard tests \`abs(\${1.51})\` 1`] = `"abs(1.51)"`;

exports[`standard tests \`abs(\${1.501})\` 1`] = `"abs(1.501)"`;

exports[`standard tests \`abs(-10)\` 1`] = `"abs((-10))"`;

exports[`standard tests \`ceil(\${-1.5})\` 1`] = `"ceil(-1.5)"`;

exports[`standard tests \`ceil(\${-1.49})\` 1`] = `"ceil(-1.49)"`;

exports[`standard tests \`ceil(\${-1.51})\` 1`] = `"ceil(-1.51)"`;

exports[`standard tests \`ceil(\${-1.501})\` 1`] = `"ceil(-1.501)"`;

exports[`standard tests \`ceil(\${1.5})\` 1`] = `"ceil(1.5)"`;

exports[`standard tests \`ceil(\${1.49})\` 1`] = `"ceil(1.49)"`;

exports[`standard tests \`ceil(\${1.51})\` 1`] = `"ceil(1.51)"`;

exports[`standard tests \`ceil(\${1.501})\` 1`] = `"ceil(1.501)"`;

exports[`standard tests \`floor(\${-1.5})\` 1`] = `"floor(-1.5)"`;

exports[`standard tests \`floor(\${-1.49})\` 1`] = `"floor(-1.49)"`;

exports[`standard tests \`floor(\${-1.51})\` 1`] = `"floor(-1.51)"`;

exports[`standard tests \`floor(\${-1.501})\` 1`] = `"floor(-1.501)"`;

exports[`standard tests \`floor(\${1.5})\` 1`] = `"floor(1.5)"`;

exports[`standard tests \`floor(\${1.49})\` 1`] = `"floor(1.49)"`;

exports[`standard tests \`floor(\${1.51})\` 1`] = `"floor(1.51)"`;

exports[`standard tests \`floor(\${1.501})\` 1`] = `"floor(1.501)"`;

exports[`standard tests \`foo() + foo(1) + foo(1, 2) + foo(1, 2, 3)\` 1`] = `"(((foo() + foo(1)) + foo(1, 2)) + foo(1, 2, 3))"`;

exports[`standard tests \`round(\${-1.5})\` 1`] = `"round(-1.5)"`;

exports[`standard tests \`round(\${-1.49})\` 1`] = `"round(-1.49)"`;

exports[`standard tests \`round(\${-1.51})\` 1`] = `"round(-1.51)"`;

exports[`standard tests \`round(\${-1.501})\` 1`] = `"round(-1.501)"`;

exports[`standard tests \`round(\${1.5})\` 1`] = `"round(1.5)"`;

exports[`standard tests \`round(\${1.49})\` 1`] = `"round(1.49)"`;

exports[`standard tests \`round(\${1.51})\` 1`] = `"round(1.51)"`;

exports[`standard tests \`round(\${1.501})\` 1`] = `"round(1.501)"`;

exports[`standard tests \`trunc(\${-1.5})\` 1`] = `"trunc(-1.5)"`;

exports[`standard tests \`trunc(\${-1.49})\` 1`] = `"trunc(-1.49)"`;

exports[`standard tests \`trunc(\${-1.51})\` 1`] = `"trunc(-1.51)"`;

exports[`standard tests \`trunc(\${-1.501})\` 1`] = `"trunc(-1.501)"`;

exports[`standard tests \`trunc(\${1.5})\` 1`] = `"trunc(1.5)"`;

exports[`standard tests \`trunc(\${1.49})\` 1`] = `"trunc(1.49)"`;

exports[`standard tests \`trunc(\${1.51})\` 1`] = `"trunc(1.51)"`;

exports[`standard tests \`trunc(\${1.501})\` 1`] = `"trunc(1.501)"`;

exports[`standard tests \`trunc(300*1.08)\` 1`] = `"trunc((300 * 1.08))"`;

exports[`standard tests \`trunc(320*1.08)\` 1`] = `"trunc((320 * 1.08))"`;
